src/client/dfs/dfs_sys.c

```diff
@@ -0,0 +1,1315 @@
+		if (rc != 0) {
+			D_DEBUG(DB_TRACE, "dfs_lookup_rel() %s failed: (%d)\n",
+				sys_path.name, rc);
+			D_GOTO(out_free_path, rc);
+		}
+
+		dfs_release(obj);
+
+		/* A link itself is always accessible */
+		if (S_ISLNK(mode))
+			D_GOTO(out_free_path, rc);
+	}
+
+	/* Either we are following symlinks, the obj is root,
+	 * or the obj is not a symlink. So just call dfs_access.
+	 */
+	rc = dfs_access(dfs_sys->dfs, sys_path.parent, sys_path.name, mask);
+
+out_free_path:
+	sys_path_free(dfs_sys, &sys_path);
+	return rc;
+}
+
+int
+dfs_sys_chmod(dfs_sys_t *dfs_sys, const char *path, mode_t mode)
+{
+	int		rc;
+	struct sys_path	sys_path;
+
+	if (dfs_sys == NULL)
+		return EINVAL;
+	if (path == NULL)
+		return EINVAL;
+
+	rc = sys_path_parse(dfs_sys, &sys_path, path);
+	if (rc != 0)
+		return rc;
+
+	rc = dfs_chmod(dfs_sys->dfs, sys_path.parent, sys_path.name, mode);
+
+	sys_path_free(dfs_sys, &sys_path);
+
+	return rc;
+}
+
+int
+dfs_sys_setattr(dfs_sys_t *dfs_sys, const char *path, struct stat *stbuf,
+		int flags, int sflags)
+{
+	int		rc;
+	struct sys_path	sys_path;
+	dfs_obj_t	*obj;
+	int		lookup_flags = O_RDWR;
+
+	if (dfs_sys == NULL)
+		return EINVAL;
+	if (path == NULL)
+		return EINVAL;
+
+	rc = sys_path_parse(dfs_sys, &sys_path, path);
+	if (rc != 0)
+		return rc;
+
+	/* No need to lookup the root */
+	if (sys_path.name == NULL) {
+		obj = sys_path.parent;
+		goto setattr;
+	}
+
+	if (sflags & O_NOFOLLOW)
+		lookup_flags |= O_NOFOLLOW;
+
+	rc = dfs_lookup_rel(dfs_sys->dfs, sys_path.parent,
+			    sys_path.name, lookup_flags, &obj, NULL, NULL);
+	if (rc != 0) {
+		D_DEBUG(DB_TRACE, "dfs_lookup_rel() %s failed: (%d)\n",
+			sys_path.name, rc);
+		D_GOTO(out_free_path, rc);
+	}
+
+setattr:
+	rc = dfs_osetattr(dfs_sys->dfs, obj, stbuf, flags);
+	if (rc != 0) {
+		D_DEBUG(DB_TRACE, "dfs_osetattr() %s failed: (%d)\n",
+			sys_path.name, rc);
+		D_GOTO(out_close_obj, rc);
+	}
+
+out_close_obj:
+	if (sys_path.name != NULL)
+		dfs_release(obj);
+out_free_path:
+	sys_path_free(dfs_sys, &sys_path);
+	return rc;
+}
+
+int
+dfs_sys_utimens(dfs_sys_t *dfs_sys, const char *path,
+		const struct timespec times[2], int flags)
+{
+	int		rc;
+	struct stat	stbuf;
+
+	stbuf.st_atim = times[0];
+	stbuf.st_mtim = times[1];
+
+	rc = dfs_sys_setattr(dfs_sys, path, &stbuf,
+			     DFS_SET_ATTR_ATIME | DFS_SET_ATTR_MTIME,
+			     flags);
+	return rc;
+}
+
+int
+dfs_sys_stat(dfs_sys_t *dfs_sys, const char *path, struct stat *buf,
+	     int flags)
+{
+	int		rc;
+	struct sys_path	sys_path;
+	dfs_obj_t	*obj;
+	int		lookup_flags = O_RDWR;
+
+	if (dfs_sys == NULL)
+		return EINVAL;
+	if (path == NULL)
+		return EINVAL;
+
+	rc = sys_path_parse(dfs_sys, &sys_path, path);
+	if (rc != 0)
+		return rc;
+
+	/* No need to lookup root */
+	if (sys_path.name == NULL) {
+		rc = dfs_ostat(dfs_sys->dfs, sys_path.parent, buf);
+		D_GOTO(out_free_path, rc);
+	}
+
+	if (flags & O_NOFOLLOW)
+		lookup_flags |= O_NOFOLLOW;
+
+	rc = dfs_lookup_rel(dfs_sys->dfs, sys_path.parent,
```

Comment:
Would dfs_stat() work here?

Reply:
dfs_stat() performs like lstat() (it interrogates the link itself).
So we could do this:
```
if (flags & O_NOFOLLOW)
    dfs_stat(...) // Does NOT follow links
else
    dfs_lookup_rel(...) Follows links by default
```
I went back and forth on this, and decided that just calling dfs_lookup_rel() all the time was simpler.
But... maybe dfs_stat() is more efficient in the case of O_NOFOLLOW.
What do you think?